File: Solutions/104-Maximum Depth of Binary Tree/104-Maximum Depth of Binary Tree.go
//Second fastest
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	return max(maxDepth(root.Left), maxDepth(root.Right)) + 1
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

//Fastest
func maxDepth(root *TreeNode) int {
    
    if root == nil {
        return 0
    }
    
    m := 0
    if root.Left == nil && root.Right == nil {
        m = 0
    } else if root.Left != nil && root.Right != nil {
        m = max(maxDepth(root.Left), maxDepth(root.Right))
    } else if root.Left != nil {
        m = maxDepth(root.Left)
    } else {
        m = maxDepth(root.Right)
    }

    return m + 1
    
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: Solutions/104-Maximum Depth of Binary Tree/104-Maximum Depth of Binary Tree.cpp
//Second fastest
class Solution {
public:
    int maxDepth(TreeNode* root) {        
        //return root == NULL ? 0:  1 + max(maxDepth(root->left), maxDepth(root->right)); // ok
		return root ? 1 + max(maxDepth(root->left), maxDepth(root->right)) : 0;
    }
};

//Another
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if(root==NULL)
            return 0;
return max(maxDepth(root->left),maxDepth(root->right))+1;
       
    }
};

//Fastest
class Solution {
    
public:
    int findMax (TreeNode* n,int depth){
        if (n == nullptr) return depth;
        
        return max(findMax(n->left, depth +1), findMax(n->right, depth +1));
        
    }
    int maxDepth(TreeNode* root) {
        return findMax ( root,0);
    }
};